docs: WIP (#133)
jet/dotnet-templates: README.md

      $ dotnet new -u Equinox.Templates

<a name="guidance"></a>
# PATTERNS / GUIDANCE

<a name="tldr"></a>
## TL;DR

1. DO have global strongly typed id types in `namespace Domain`
2. DONT have global `module Types`. AVOID per Aggregate `module Types` or top level `type` definitions

> 3. DO group stuff predictably per `module Aggregate`: `Stream, Events, Reactions, Fold, Decisions, Service, Factory`. And keep grouping within that.

Reviewer: ```suggestion
3. DO group stuff predictably per `module Aggregate`: `Stream, Events, Reactions, Fold, Decide, Service, Factory`. And keep grouping within that.
```

> 4. DONT `open <Aggregate>`, `open <Aggregate>.Events`, `open <Aggregate>.Fold`

Reviewer: counterpoint

```fsharp
module Fold = 
    open Events
```

Reviewer: also useful to `open Aggregate` in test files, but can do without

Author: I didnt call those out further down but I will. I also intend to link these tl;dr sentences to sections below

Author: Hopefully https://github.com/jet/dotnet-templates/pull/133/commits/c1a7ff54dbe288b5f3eeaa6c4b2fbe9e2ffc0e23 covers it

5. DO design for idempotency everywhere
6. DONT use `Result` or a per-Aggregate `type Error`. DO use minimal result types per decision function
7. DONT pass out `Fold.State` from a `Service`

> 8. DONT be a slave to CQRS for all read paths. CONSIDER AllowState. AVOID Query

Reviewer: Can you clarify `AVOID Query`? I have query modules converting `Fold.State -> Query.Model`

Reviewer: ```suggestion
8. DONT be a slave to CQRS for all read paths. CONSIDER AllowStale. AVOID Query
```

Author: And yes, I had seen that typo!

Author: Hopefully covered by https://github.com/jet/dotnet-templates/pull/133/commits/c5c9aa48fe4b1bd9d00dc80022dbc55b1ee7cfcd

> 9. DONT be a slave to the Command pattern

Reviewer: We might as well go all out and discourage the pattern entirely. The way Equinox code is written the command pattern, even in its simplest state, results in more harder to read code.

```fsharp
module Decide =
    type Command = Increment | Decrement
    let decide command state =
        match command with
        | Increment -> if state > 10 then [] else [Events.Incremented]
        | Decrement -> if state = 0 then [] else [Events.Decremented]
```

vs

```fsharp
module Decide =
    let increment state = if state > 10 then [] else [Events.Incremented]
    let decrement state = if state = 0 then [] else [Events.Decremented]
```

Author: yes, or `let increment state = if state > 10 then [] else [Events.Incremented]` => `let increment state = [if state <= 10 then Events.Incremented]`

Author: attempted to cover this in [283727e](https://github.com/jet/dotnet-templates/pull/133/commits/283727e82cc717f5e62093d0c373ab730edc9979) but not really satisfied as I went too far. I hate the Counter example!

      $ dotnet new -u Equinox.Templates

> <a name="guidance"></a>

Author: @nordfjord is patterns a better stable anchor than guidance? i.e. Are these actually patterns? antipatterns? pushy guidance?

Reviewer: I don't think these count as patterns. I think we're in pushy guidance territory

...

// <Aggregate>.fs
module Aggregate

open Domain.Types

module Events =

    type Event =
        | Created of {| creator: UserId; context: EntityContext |}
        ...

// <Aggregate2>.fs
module Aggregate2

module Events =

    type Event =
        | Copied of {| by: UserId; context: Types.EntityContext |}
        ...
```

Instead:
- have each `module <Aggregate>` have its own version of each type that will be used in an event _within its `module Events`_. (The `decide` function can map from an input type if desired, but the important thing is that the Aggregate will need to be able to roundtrip its types in perpetuity, and having to disentangle the overlap between more than on Aggregate is simply never a good tradeoff) 
- [sharing id types is fine](#global-do-share-ids)

<a name="global-do-share-ids"></a>
### ✅ DO have global strongly typed ids

While [sharing the actual types is a no-no](#global-dont-share-types), having common id types is perfectly reasonable

```fsharp
module Domain.Types

type UserId = ...
type TenantId = ...

...

module Domain.User

module Events =

    type Joined = { tenant: TenantId; authorizedBy: UserId }

```

> - CONSIDER UMX for non-serialized ids

Reviewer: what does it mean for an ID to be `non-serialised`?

Author: I don't round-trip the ids in the equinox (or dotnet-templates) samples to a store, so:
- serialization wrt FsCodec bodies doesn't matter
- xss and model binding matter more than they would as they are used in dtos

Wherever possible, the templates use use strongly type identifiers, particularly ones that might naturally be represented as primitives, i.e. `string` etc.

- [`FSharp.UMX`](https://github.com/fsprojects/FSharp.UMX) is useful to transparently pin types in a message contract cheaply - it works well for a number of contexts:

    - Coding/decoding events using [FsCodec](https://github.com/jet/fscodec). (because Events are things that **have happened**, validating them is not a central concern as we load and fold these incontrovertible Facts)
    - Model binding in ASP.NET (because the types de-sugar to the primitives, no special support is required). _Unlike events, there are more considerations in play in this context though; often you'll want to apply validation to the inputs (representing Commands) as you map them to [Value Objects](https://martinfowler.com/bliki/ValueObject.html), [Making Illegal States Unrepresentable](https://fsharpforfunandprofit.com/posts/designing-with-types-making-illegal-states-unrepresentable/). Often, Single Case Discriminated Unions can be a better tool inb that context_

> ### ✅ DO Have a helper `module` per id type

Reviewer: ```fsharp
type UserId = Guid<userId>
and [<Measure>] userId

module UserId =
    let toGuid (id: UserId): Guid = %id
    let toString (id: UserId): string = (toGuid id).ToString("N")
    let ofGuid (id: Guid): UserId = %id 
    let parse (id: string): UserId = ofGuid (Guid.Parse id)
```

Author: ty sir https://github.com/jet/dotnet-templates/pull/133/commits/5519c3572ebb9a93417a0c1ebec51507f4eb507e

#### ❌ DONT split the `module <Aggregate>`
Having the Event Contracts, State and Decision logic in a single module can feel wrong when you get over e.g. 1000 lines of code; instincts to split the file on some basis will kick in. Don't do it; splitting the file is hiding complexity under the carpet.

#### ❌ DONT move the `module Events` out

The Event Contracts are the most important contract that an Aggregate has - decision logic will churn endlessly. You might even implement logic against it in other languages. But the Event Contracts you define are permanent. As a developer fresh to a project, the event contracts are often the best starting point as you try to understand what a given aggregate is responsible for. 

#### ❌ DONT move the `module State`, or `evolve` logic out

The State type and the associated `evolve` and `fold` functions are intimately tied to the Event Contracts. Over time, ugliness and upconversion can lead to noise, and temptation to move it out. Don't do it; being able to understand the full coupling is critical to understanding how things work, and equally critical to being able to change or add functions.

> #### ❌ DONT move the decision logic out

Reviewer: ![image](https://github.com/jet/dotnet-templates/assets/3721521/c90f4934-461e-4331-99e3-866ad9b6ea27)

Author: it wasn't quite dead IMO https://github.com/jet/dotnet-templates/pull/133/commits/4ba2cae305fa7474f46c4969fdbd339379358e73

### 2. `module Events`

Having the Event Contracts be their own `module` is a critical forcing function for good aggregate design. Having all types and all cases live in one place and being able to quickly determine where each Event is produced is key to being able to understand the moving parts of a system.

<a name="events-no-ids"></a>
#### ❌ AVOID including egregious identity information

When modelling, it's common to include primary identifiers (e.g. a user id), or contextual identifiers (e.g. a tenant id) in an Event in order to convey the relationships between events in the systems as a whole; you want the correlations to stand out. In the implementation however, repeating the identity information in every event is a major liability:
1. the State needs to contain the values - that's more noise
2. Event versioning gets messier - imagine extending a system to make it multi-tenant, you'd need to be able to handle all the historic events that predated the concept

> The alternative is for a workflow to react to the events in the context of a stream - if some logic needs to know the userid let the User reactor handing the User event on a User Stream pass thgat context forward if relevant in that context.

Reviewer: typos
```suggestion
The alternative is for a workflow to react to the events in the context of a stream - if some logic needs to know the userid let the `User` reactor handling the `User` event on a `User` Stream pass that context forward if relevant in that context.
```

"pass that context forward if relevant in that context" <- Yo dawg, I heard you like context, so I put a context in your context so you can context while you context

#### ❌ DONT `open Events` in an aggregate module

Having to prefix types and/or Event Type names with `Events.` is a feature, not a bug. 

### 3. `module Fold`

<a name="fold-no-log"></a>
#### ❌ DONT log

> If Fold logic is not incredibly boring, that's a design smell. In general, you want to [make illegal States unrepresentable](https://fsharpforfunandprofit.com/posts/designing-with-types-making-illegal-states-unrepresentable/). If you must, unit test it to satisfy yourself things can't go wrong.

Reviewer: > If Fold logic is not incredibly boring, that's a design smell

Reword, "If Fold logic is complicated" or "If Fold logic is anything but incredibly boring"

Took a while to parse due to the double negative

Author: https://github.com/jet/dotnet-templates/pull/133/commits/87a764c2e18a7a62c693a020f806d1c9b0af0f3a
